authenticate-users/index.php

```diff
@@ -0,0 +1,119 @@
+ * Returns a dictionary of current Google public key certificates for
+ * validating Google-signed JWTs. Since these change rarely, the result
+ * is cached on first request for faster subsequent responses.
+ */
+function certs() : string
+{
+    $client = new Guzzle\Http\Client();
+    $response = $client->get(
+        'https://www.gstatic.com/iap/verify/public_key-jwk'
+    )->send();
+    return $response->getBody();
+}
+# [END getting_started_auth_certs]
+
+# [START getting_started_auth_metadata]
+/**
+ * Returns a string with the project metadata value for the item_name.
+ * See https://cloud.google.com/compute/docs/storing-retrieving-metadata for
+ * possible item_name values.
+ */
+function get_metadata($itemName) : string
+{
+    $client = new Guzzle\Http\Client();
+
+    $endpoint = 'http://metadata.google.internal';
+    $path = '/computeMetadata/v1/project/' . $itemName;
+    $response = $client->get(
+        $endpoint . $path,
+        null,
+        ['headers' => ['Metadata-Flavor' => 'Google']]
+    )->send();
+
+    return $response->getBody();
+}
+# [END getting_started_auth_metadata]
+
+# [START getting_started_auth_audience]
+/**
+ * Returns the audience value (the JWT 'aud' property) for the current
+ * running instance. Since this involves a metadata lookup, the result is
+ * cached when first requested for faster future responses.
+ */
+function audience() : string
+{
+    $projectNumber = get_metadata('numeric-project-id');
+    $projectId = get_metadata('project-id');
+    $audience = sprintf('/projects/%s/apps/%s', $projectNumber, $projectId);
+    return $audience;
+}
+# [END getting_started_auth_audience]
+
+# [START getting_started_auth_validate_assertion]
+/**
+ * Checks that the JWT assertion is valid (properly signed, for the
+ * correct audience) and if so, returns strings for the requesting user's
+ * email and a persistent user ID. If not valid, returns None for each field.
+ */
+function validate_assertion($assertion) : array
+{
+    $jwkset = new SimpleJWT\Keys\KeySet();
+    $jwkset->load(certs());
+    try {
+        $info = SimpleJWT\JWT::decode(
+            $assertion,
+            $jwkset,
+            'ES256'
+        );
+        if ($info->getClaim('aud') != audience()) {
+            throw new Exception('Audience did not match');
+        }
+        return [$info->getClaim('email'), $info->getClaim('sub')];
+    } catch (Exception $e) {
+        printf('Failed to validate assertion: %s', $e->getMessage());
+        return [null, null];
+    }
+}
+# [END getting_started_auth_validate_assertion]
+
+# [START getting_started_auth_front_controller]
+/**
+ * This is an example of a front controller for a flat file PHP site. Using a
+ * Static list provides security against URL injection by default.
+ */
+switch (@parse_url($_SERVER['REQUEST_URI'])['path']) {
+    case '/':
+        $assertion = getallheaders()['X-Goog-Iap-Jwt-Assertion'] ?? '';
+        list($email, $id) = validate_assertion($assertion);
```

Comment:
- Should we be doing something with `$id`?
- If email/id come back null, why also output `Hello None` instead of just letting the above `Failed to validate..` message be all that is shown?

Reply:
I think it's just to let the user know it's available. I think it's ok to do nothing with it, personally.

Agreed, I added a conditional so "Hello" is only displayed if $email is not empty.